core/modules/loader/DoListItem.h

```diff
@@ -0,0 +1,180 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2018 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ *
+ */
+#ifndef LSST_QSERV_LOADER_DOLISTITEM_H
+#define LSST_QSERV_LOADER_DOLISTITEM_H
+
+// system headers
+#include <chrono>
+#include <list>
+
+// Qserv headers
+#include "util/ThreadPool.h"
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+
+class Central;
+
+
+class TimeOut {
+public:
+    using TimePoint = std::chrono::system_clock::time_point;
+    using Clock = std::chrono::system_clock;
+
+    explicit TimeOut(std::chrono::milliseconds timeOut) : _timeOut(timeOut) {}
+
+    bool due() { return due(Clock::now()); }
+    bool due(TimePoint now) {
+        auto triggerDiff = std::chrono::duration_cast<std::chrono::milliseconds>(now - _lastTrigger);
+        return (triggerDiff > _timeOut);
+    }
+
+    void triggered() { return triggered(Clock::now()); }
+    void triggered(TimePoint now) {
+        _lastTrigger = now;
+    }
+
+    std::chrono::milliseconds timeLeft(TimePoint now) {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(now - _lastTrigger);
+    }
+
+    void setTimeOut(std::chrono::milliseconds timeOut) { _timeOut = timeOut; }
+    std::chrono::milliseconds getTimeOut() const { return _timeOut; }
+private:
+    // How much time since lastTrigger needs to pass before triggering.
+    std::chrono::milliseconds _timeOut; // default, 15 minutes
+    TimePoint _lastTrigger{std::chrono::seconds(0)};
+};
+
+
+/// Children of this class *MUST* be created with shared pointers.
```

Comment:
I think you mean 'instances' instead of 'children'? The latter implies (IMHO) a 'has' relationship; a parent has a child. The former is a thing that is created; a "DoListItem instance must be managed by a shared pointer".

Reply:
DoListItem is a base class for other classes, and those classes must be instantiated with shared pointers or bad things will happen. I'll edit the comments.